### PR TITLE
OSASINFRA-3179: openstack failureDomain: Add RootVolume type

### PR DESCRIPTION
**Pull request by pierreprinetti**
The volumeType is as of OpenStack v17 the canonical way to distribute storage in separate failure domains. This change introduces the ability to set a variable volume type for the Control plane instances.

/cc mandre emilienm

### CODE REVIEW

**Comment by EmilienM** [comment]
Nice addition, thanks. Could you add a test to the testsuite among rootVolume.availabilityZone maybe? Thanks

File: machine/v1/types_controlplanemachineset.go [hunk 1]
```diff
@@ -340,6 +340,15 @@ type RootVolume struct {
...
 	AvailabilityZone string `json:"availabilityZone,omitempty"`
+
+	// volumeType specifies the type of the root volume that will be provisioned.
+	// If not specifified, the root volume will be created as the type in the machine template.
+	// The maximum length of a volume type name is 255 as per the OpenStack limit.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=255
+	// +kubebuilder:validation:Pattern=`^[^ ]*$`
+	// +optional
+	VolumeType string `json:"volumeType,omitempty"`
```
Comment (EmilienM) [inline machine/v1/types_controlplanemachineset.go:351]: This corresponds exactly with our machinev1alpha1 API:
https://github.com/openshift/api/blob/master/machine/v1alpha1/types_openstack.go#L358

File: machine/v1/types_controlplanemachineset.go [hunk 2]
```diff
@@ -340,6 +340,15 @@ type RootVolume struct {
 	// +kubebuilder:validation:Pattern=`^[^ ]*$`
 	// +optional
 	AvailabilityZone string `json:"availabilityZone,omitempty"`
+
+	// volumeType specifies the type of the root volume that will be provisioned.
+	// If not specifified, the root volume will be created as the type in the machine template.
+	// The maximum length of a volume type name is 255 as per the OpenStack limit.
```
Comment (JoelSpeed) [inline machine/v1/types_controlplanemachineset.go:346]: Nit
```suggestion
	// The maximum length of a volume type name is 255 characters, as per the OpenStack limit.
```

**Comment by pierreprinetti** [comment]
/hold cancel
This should be ready for another pass!

File: machine/v1/types_controlplanemachineset.go [hunk 3]
```diff
@@ -340,6 +340,14 @@ type RootVolume struct {
...
 	// +optional
 	AvailabilityZone string `json:"availabilityZone,omitempty"`
+
+	// volumeType specifies the type of the root volume that will be provisioned.
+	// If not specifified, the root volume will be created as the type in the machine template.
+	// The maximum length of a volume type name is 255 characters, as per the OpenStack limit.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=255
+	// +optional
+	VolumeType string `json:"volumeType,omitempty"`
```
Comment (JoelSpeed) [inline machine/v1/types_controlplanemachineset.go:350]: Are there definitely no character limitations on this? Are there any examples of valid volume types we can include in the godoc to give the user an idea of a correct value?

Comment (JoelSpeed) [reply machine/v1/types_controlplanemachineset.go:350]: Ok, in that case I guess there's not a lot we can do

**Comment by JoelSpeed** [comment]
Are we expecting any further changes? Has this been tested at all to give us confidence in the API? I have no further feedback given we've gone through and established there's no patterns etc that can be applied

**Comment by pierreprinetti** [comment]
I don't expect any further changes on this. As for the pattern, I have slavishly followed the OpenStack ProviderSpec nomenclature.

Tests will happen in the [CPMS PR](https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/217).

**Pull request merged**